src/muya/lib/selection/cursor.js

```diff
@@ -0,0 +1,43 @@
+import { compareParagraphsOrder } from './dom'
+
+class Cursor {
+  // You need to provide either `anchor`&&`focus` or `start`&&`end` or all.
+  constructor ({ anchor, focus, start, end, noHistory = false }) {
+    if (anchor && focus && start && end) {
+      this.anchor = anchor
+      this.focus = focus
+      this.start = start
+      this.end = end
+    } else if (anchor && focus) {
+      this.anchor = anchor
+      this.focus = focus
+      if (anchor.key === focus.key) {
+        if (anchor.offset <= focus.offset) {
+          this.start = this.anchor
+          this.end = this.focus
+        } else {
+          this.start = this.focus
+          this.end = this.anchor
+        }
+      } else {
+        const anchorParagraph = document.querySelector(`#${anchor.key}`)
+        const focusParagraph = document.querySelector(`#${focus.key}`)
+        const order = compareParagraphsOrder(anchorParagraph, focusParagraph)
+
+        if (order) {
+          this.start = this.anchor
+          this.end = this.focus
+        } else {
+          this.start = this.focus
+          this.end = this.anchor
+        }
+      }
+    } else {
```

Comment:
You can improve this by throwing an exception if `start` and `end` are null. E.g. adding an `else if (start && end) ... else throw error` block.

Reply:
Cursor is support has `start is null` parameter, because in getCursorRange method, start or end may be `null`, we need to check if it is null in the conentState folder.